core/state_transition.go

```diff
@@ -754,13 +795,17 @@ func (st *StateTransition) refundGas(refundQuotient uint64) uint64 {
 	st.gasRemaining += refund
 
 	// Return ETH for remaining gas, exchanged at the original rate.
-	if st.gasFromSoul {
-		remaining := new(big.Int).Mul(new(big.Int).SetUint64(st.gasRemaining), st.msg.GasPrice)
-		st.AddSoulBalance(st.msg.From, remaining, tracing.BalanceIncreaseGasReturn)
-	} else {
-		remaining := uint256.NewInt(st.gasRemaining)
-		remaining.Mul(remaining, uint256.MustFromBig(st.msg.GasPrice))
+	remaining := uint256.NewInt(st.gasRemaining)
+	remaining.Mul(remaining, uint256.MustFromBig(st.msg.GasPrice))
+	if st.usedSGTBalance == nil {
 		st.state.AddBalance(st.msg.From, remaining, tracing.BalanceIncreaseGasReturn)
+	} else {
+		if remaining.Cmp(st.usedSGTBalance) <= 0 {
+			st.addSoulBalance(st.msg.From, remaining, tracing.BalanceIncreaseGasReturn)
```

Comment:
should the remaining be refunded to the native balance first, and then the soul balance?

Reply:
Done in https://github.com/ethstorage/op-geth/pull/16/commits/6b256670dd65445bf1e6d9d15794affddeac159d .